Update endpoint

Aurora1024 opened the pull request:
Test Cases for YourResourceModel Model 
- Test create a recommendation
- Test deserialization of a Recommendation
- Test deserialization of bad data
- Test deserialization of a Recommendation with missing data
- Find a recommendation type by two product ids
- Test serialization of a Recommendation
- Update a recommendation type by two product ids
- Update a recommendation type by two product ids

REST API Recommendation Server Tests 
- Create a new recommendation
- Create a Recommendation with missing data
- Create a Recommendation with no content type
- Get a single Recommendation
- Get a Recommendation thats not found
- Test index call
- Update an existing recommendation
- create a Recommendation with no content type
- update a Recommendation that is not found


Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
service/__init__.py      25      4    84%   30, 39-42
service/models.py        55      0   100%
service/routes.py        52      0   100%
service/status.py        46      0   100%
---------------------------------------------------
TOTAL                   178      4    98%
-----------------------------------------------------------------------------
17 tests run in 0.905 seconds (17 tests passed)

Aurora1024 committed:
update

Aurora1024 committed:
basic done

14 tests run in 0.805 seconds (14 tests passed)

zyy9871 commented on service/routes.py:
```diff
@@ -80,6 +80,33 @@ def get_recommendations(product_id1, product_id2):
+    recommendation.deserialize(request.get_json())
+    old_recommendation = recommendation.find(product_id1, product_id2)
+    if not old_recommendation:
+        raise BadRequest("No existing relationship between product {} and {}".format(product_id1, product_id2))
```
Should raise NotFound. No idea how to cover this line.

mandy-cmd replied:
To cover this line you could write a unit test trying to update relationship between -1 and -1 and a NOT FOUND as mentioned above should be expected in your test.

Aurora1024 replied:
done in test_route

zyy9871 commented on service/routes.py:
```diff
@@ -80,6 +80,33 @@ def get_recommendations(product_id1, product_id2):
+    elif recommendation.relationship == "UP_SELL":
+        old_recommendation.relationship = Type.UP_SELL
+    elif recommendation.relationship == "ACCESSORY":
+        old_recommendation.relationship = Type.ACCESSORY
```
Just do old_recommendation.relationship = recommendation.relationship

mandy-cmd replied:
Yes old_recommendation.relationship = recommendation.relationship should be sufficient. Also here's also where you are missing coverage but I don't think you have to test every type UPDATE. Making it old_recommendation.relationship = recommendation.relationship should also help with testing coverage.

mandy-cmd commented on service/models.py:
```diff
@@ -51,6 +51,15 @@ def create(self):
+        """
+        update a recommendation type to the database
+        """
+        if not self.relationship:
```
To test this line, have a test case say test_update_empty_relationship() and do not set relationship in your input. Then DataValidationError should be expected.

Aurora1024 replied:
Could you pls help look at test_update_recommendation_no_relation_type(self)
I try to assign null to relation type in json file, but it raises error when deserialize it. A little confused how to test this one

 service/routes.py line 92 in update_recommendations
      recommendation.deserialize(request.get_json())
    service/models.py line 83 in deserialize
      raise DataValidationError(
   DataValidationError: Invalid YourResourceModel: body of request contained bad or no data
   -------------------- >> begin captured logging << --------------------
   root: DEBUG: <enum 'Type'>
   root: DEBUG: {'product_id1': 1, 'product_id2': 7, 'relationship': 'CROSS_SELL'}
   root: DEBUG: {"product_id1": 1, "product_id2": 7, "relationship": " "}
   --------------------- >> end captured logging << ---------------------

mandy-cmd commented on tests/test_routes.py:
```diff
@@ -128,3 +128,28 @@ def test_create_recommendation_no_content_type(self):
+            json=new_recommendation,
+            content_type="application/json"
+        )
+       #new_recommendation["relationship"] = "CROSS_SELL"
```
Please remove commented/unused code.

mandy-cmd commented on tests/test_models.py:
```diff
@@ -106,4 +106,21 @@ def test_find_recommendation_type(self):
+        self.assertEqual(recommendations[0].product_id1, recommendation.product_id1)
+        self.assertEqual(recommendations[0].product_id2, recommendation.product_id2)
+        self.assertEqual(recommendations[0].relationship, recommendation.relationship)
+
```
To cover DataValidationError, you could create a new test and say here relationship = None, and then dataValidationError should be thrown.

Aurora1024 replied:
Done in test_model thanks

Aurora1024 committed:
need help with test_update_recommendation_no_relation_type

pls look at test_routes line 171
test_update_recommendation_no_relation_type(self)

Aurora1024 committed:
Done

Aurora1024 committed:
delete useless code

mandy-cmd commented on tests/test_routes.py:
```diff
@@ -148,8 +149,35 @@ def test_update_recommendation(self):
+            json=test,
+            content_type="application/json"
+        )
+        self.assertEqual(resp.status_code, status.HTTP_415_UNSUPPORTED_MEDIA_TYPE)
```
Ok I see your error. So it's actually the [deserialization error](https://github.com/dev-ops-recommendation/recommendations/blob/main/service/models.py#L83) meaning that input relationship cannot be empty at all, in which case you could remove this test and remove ```if not self.relationship``` because it won't be able to hit here at all because validation error would raise first before it could reach this check. 
```
DataValidationError: Invalid YourResourceModel: body of request contained bad or no data
   -------------------- >> begin captured logging << --------------------
   root: DEBUG: {'product_id1': 3, 'product_id2': 8, 'relationship': 'CROSS_SELL'}
   root: DEBUG: {"product_id1": 3, "product_id2": 8, "relationship": " "}
   --------------------- >> end captured logging << ---------------------
```

Aurora1024 replied:
Yeah. I test this in test_model part. "if not recommendation" is implement in model.py

mandy-cmd commented on service/models.py:
```diff
@@ -51,6 +51,15 @@ def create(self):
+        """
+        update a recommendation type to the database
+        """
+        if not self.relationship:
+            raise DataValidationError("Update called with empty relationship")
```
What I mean is that I think you could remove these two lines and remove no relation tests.

Aurora1024 replied:
Right, it makes sense. But I think I can leave here for double check.

mandy-cmd replied:
Then you want to test it in test_model.py because route will never hit here if input relationship is empty.

Aurora1024 replied:
already commit new version that test it in  test_model.py . Thanks

mandy-cmd committed:
Merge branch 'main' into update
